java/src/main/java/ai/rapids/cudf/DType.java

```diff
@@ -15,78 +15,292 @@
+
+  /* This is used in fromNative method to return singleton object for non-decimal types.
+  Please make sure the order here is same as that of DTypeEnum. Whenever a new non-decimal
+  type is added in DTypeEnum, this array needs to be updated as well.*/
+  private static final DType[] SINGLETON_DTYPE_LOOKUP = new DType[]{
+      EMPTY,
+      INT8,
+      INT16,
+      INT32,
+      INT64,
+      UINT8,
+      UINT16,
+      UINT32,
+      UINT64,
+      FLOAT32,
+      FLOAT64,
+      BOOL8,
+      TIMESTAMP_DAYS,
+      TIMESTAMP_SECONDS,
+      TIMESTAMP_MILLISECONDS,
+      TIMESTAMP_MICROSECONDS,
+      TIMESTAMP_NANOSECONDS,
+      DURATION_DAYS,
+      DURATION_SECONDS,
+      DURATION_MILLISECONDS,
+      DURATION_MICROSECONDS,
+      DURATION_NANOSECONDS,
+      null, // DICTIONARY32
+      STRING,
+      LIST,
+      null, // DECIMAL32
+      null, // DECIMAL64
+      STRUCT
+  };
+
   /**
-   * s since the UNIX epoch
+   * This only works for fixed width types. Variable width types like strings the value is
+   * undefined and should be ignored.
+   *
+   * @return size of type in bytes.
    */
-  TIMESTAMP_SECONDS(8, 13, "timestamp[s]"),
+  public int getSizeInBytes() { return typeId.sizeInBytes; }
+
   /**
-   * ms since the UNIX epoch
+   * Returns scale for Decimal Type
+   * @return scale
    */
-  TIMESTAMP_MILLISECONDS(8, 14, "timestamp[ms]"),
+  public int getScale() { return scale; }
+
   /**
-   * microseconds since the UNIX epoch
+   * Returns string name mapped to type.
+   * @return name corresponding to type
    */
-  TIMESTAMP_MICROSECONDS(8, 15, "timestamp[us]"),
+  public String getSimpleName() { return typeId.simpleName; }
+
   /**
-   * ns since the UNIX epoch
+   * Return enum for this DType
+   * @return DTypeEnum
    */
-  TIMESTAMP_NANOSECONDS(8, 16, "timestamp[ns]"),
+  public DTypeEnum getTypeId() {
+    return typeId;
+  }
 
-  //We currently don't have mappings for duration type to I/O files, and these
-  //simpleNames might change in future when we do
-  DURATION_DAYS(4, 17, "int32"),
-  DURATION_SECONDS(8, 18, "int64"),
-  DURATION_MILLISECONDS(8, 19, "int64"),
-  DURATION_MICROSECONDS(8, 20, "int64"),
-  DURATION_NANOSECONDS(8, 21, "int64"),
-  //DICTIONARY32(4, 22, "NO IDEA"),
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    DType type = (DType) o;
+    return scale == type.scale && typeId == type.typeId;
+  }
 
-  STRING(0, 23, "str"),
-  LIST(0, 24, "list"),
-  STRUCT(0, 27, "struct");
+  @Override
+  public int hashCode() {
+    return Objects.hash(typeId, scale);
+  }
 
-  private static final DType[] D_TYPES = DType.values();
-  final int sizeInBytes;
-  final int nativeId;
-  final String simpleName;
+  @Override
+  public String toString() {
+    if (isDecimalType()) {
+      return "" + typeId + " scale:" + scale;
+    } else {
+      return "" + typeId;
```

Comment:
nit: I find `typeId.toString` or `String.valueOf(typeId)` to be cleaner, even if it is larger.